### PR TITLE
Add guidelines for C API return values

### PR DESCRIPTION
**Pull request by erlend-aasland**
----
:books: Documentation preview :books:: https://cpython-devguide--1123.org.readthedocs.build/

### CODE REVIEW

**Comment by erlend-aasland** [comment]
I made a PR: #1123

**Review by markshannon — APPROVED** [review]
Looks good to me, as a first step.

File: developer-workflow/c-api.rst [hunk 1]
```diff
@@ -106,6 +109,24 @@ Guidelines for expanding/changing the public API:
 - Make sure the ownership rules and lifetimes of all applicable struct
   fields, arguments and return values are well defined.
 
+- Functions returning ``PyObject *`` must return a valid pointer on success
+  and ``NULL`` with an exception raised on error.
+  Most other API must return ``-1`` with an exception raised on error,
+  and ``0`` on success.
+
+- APIs with lesser and greater results must return ``0`` for the lesser result,
+  and ``1`` for the greater result.
```
Comment (vstinner) [inline developer-workflow/c-api.rst:118]: I don't undertand what "lesser" and "greater" mean here.

Comment (erlend-aasland) [reply developer-workflow/c-api.rst:118]: There is an example, just below. Is that sufficient?

File: developer-workflow/c-api.rst [hunk 2]
```diff
@@ -106,6 +109,24 @@ Guidelines for expanding/changing the public API:
 - Make sure the ownership rules and lifetimes of all applicable struct
   fields, arguments and return values are well defined.
 
+- Functions returning ``PyObject *`` must return a valid pointer on success
```
Comment (CAM-Gerlach) [inline developer-workflow/c-api.rst:112]:
```suggestion
- Functions returning ``PyObject *`` must return a valid pointer on success,
```

While the SemBr makes the intent clear in the source, I was confused reading the rendered prose due to the ambiguous syntax and actually read it as (success and ``NULL``), (with an exception raised on error). A comma here clears that up.

File: developer-workflow/c-api.rst [hunk 3]
```diff
@@ -106,6 +109,24 @@ Guidelines for expanding/changing the public API:
...
+
+  - ``return -1``: internal error or API misuse; exception raised
+  - ``return 0``: lookup succeeded; no item was found
+  - ``return 1``: lookup succeeded; item was found
+
+Please start a public discussion if these guidelines won't work for your API.
+
+.. note::
+
+   With *return value*, we mean the value returned by the *C return statement*.
```
Comment (CAM-Gerlach) [inline developer-workflow/c-api.rst:129]:
```suggestion
   By *return value*, we mean the value returned by the *C return statement*.
```

Use the appropriate word here

**Review by CAM-Gerlach — COMMENTED** [review]
A couple small textual tweaks, otherwise LGTM

**Review by CAM-Gerlach — APPROVED** [review]
LGTM, thanks @erlend-aasland !

**Comment by erlend-aasland** [comment]
Unless there are further remarks, I consider landing this early the coming week.

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by erlend-aasland** [comment]
Thanks for the reviews!

FTR, I also got a thumbs up by Thomas Wouters for this PR.

**Pull request merged**